src/webgl/p5.RendererGL.js

```diff
@@ -127,151 +144,178 @@ p5.RendererGL.prototype.background = function() {
+      var shader = gl.createShader(shaderTypes[i]);
+      gl.shaderSource(shader, flagPrefix + shaders[i]);
+      gl.compileShader(shader);
+      if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
+        console.log('Yikes! An error occurred compiling the shaders:' +
+          gl.getShaderInfoLog(shader));
+        return null;
+      }
+      gl.attachShader(shaderProgram, shader);
+    }
 
-  return shaderProgram;
-};
+    gl.linkProgram(shaderProgram);
+    if (!gl.getProgramParameter(shaderProgram, gl.LINK_STATUS)) {
+      console.log('Snap! Error linking shader program');
+    }
 
-p5.RendererGL.prototype._getLocation =
-function(shaderProgram, isImmediateMode) {
-  var gl = this.GL;
-  gl.useProgram(shaderProgram);
-  shaderProgram.uResolution =
-    gl.getUniformLocation(shaderProgram, 'uResolution');
-  gl.uniform1f(shaderProgram.uResolution, RESOLUTION);
-
-  //projection Matrix uniform
-  shaderProgram.uPMatrixUniform =
-    gl.getUniformLocation(shaderProgram, 'uProjectionMatrix');
-  //model view Matrix uniform
-  shaderProgram.uMVMatrixUniform =
-    gl.getUniformLocation(shaderProgram, 'uModelViewMatrix');
-
-  //@TODO: figure out a better way instead of if statement
-  if(isImmediateMode === undefined){
-    //normal Matrix uniform
-    shaderProgram.uNMatrixUniform =
-    gl.getUniformLocation(shaderProgram, 'uNormalMatrix');
-
-    shaderProgram.samplerUniform =
-    gl.getUniformLocation(shaderProgram, 'uSampler');
+    this.mHash[mId] = shaderProgram;
   }
+
+  return mId;
 };
 
+//////////////////////////////////////////////
+// UNIFORMS
+//////////////////////////////////////////////
+
 /**
- * Sets a shader uniform given a shaderProgram and uniform string
- * @param {String} shaderKey key to material Hash.
- * @param {String} uniform location in shader.
- * @param { Number} data data to bind uniform.  Float data type.
- * @todo currently this function sets uniform1f data.
- * Should generalize function to accept any uniform
- * data type.
+ * @param {String} name The name of the uniform.
  */
-p5.RendererGL.prototype._setUniform1f = function(shaderKey,uniform,data)
+p5.RendererGL.prototype._getUniform = function(uName)
 {
-  var gl = this.GL;
-  var shaderProgram = this.mHash[shaderKey];
-  gl.useProgram(shaderProgram);
-  shaderProgram[uniform] = gl.getUniformLocation(shaderProgram, uniform);
-  gl.uniform1f(shaderProgram[uniform], data);
-  return this;
+  return this._uniforms[uName].data;
 };
 
-p5.RendererGL.prototype._setMatrixUniforms = function(shaderKey) {
-  var gl = this.GL;
-  var shaderProgram = this.mHash[shaderKey];
-
-  gl.useProgram(shaderProgram);
+/**
+ * @param {String} name The name of the uniform.
+ * @param {any} data The data to set in the uniform. This can take many
+ *                   different forms. See p5.Shader.setUniform for full
+ *                   documentation.
+ */
+p5.RendererGL.prototype._setUniform = function()
+{
+  var args = Array.prototype.slice.call(arguments);
```

Comment:
use 
```
var args = new Array(arguments.length);
  for (var i = 0; i < args.length; ++i) {
    args[i] = arguments[i];
  }
```
instead of slice